ipam/crd: Fix spurious CiliumNode update status failures

gandro opened the pull request:
When running in CRD-based IPAM modes (Alibaba, Azure, ENI, CRD), it is
possible to observe spurious "Unable to update CiliumNode custom
resource" failures in the cilium-agent.

The full error message is as follows: "Operation cannot be fulfilled on
ciliumnodes.cilium.io <node>: the object has been modified; please apply
your changes to the latest version and try again".

It means that the Kubernetes `UpdateStatus` call has failed because the
local `ObjectMeta.ResourceVersion` of submitted CiliumNode version is
out of date. In the presence of races, this error is expected and will
resolve itself once the agent receives a more recent version of the
object with the new resource version.

However, it is possible that the resource version of a `CiliumNode`
object is bumped even though the `Spec` or `Status` of the `CiliumNode`
remains the same. This for examples happens when
`ObjectMeta.ManagedFields` is updated by the Kubernetes apiserver.

Unfortunately, `CiliumNode.DeepEqual` does _not_ consider any
`ObjectMeta` fields (including the resource version). Therefore two
objects with different resource versions are considered the same by the
`CiliumNode` watcher used by IPAM.

But to be able to successfully call `UpdateStatus` we need to know the
most recent resource version. Otherwise, `UpdateStatus` will always fail
until the `CiliumNode` object is updated externally for some reason.

Therefore, this commit modifies the logic to always store the most
recent version of the `CiliumNode` object, even if `Spec` or `Status`
has not changed.  This in turn allows `nodeStore.refreshNode` (which
invokes `UpdateStatus`) to always work on the most recently observed
resource version.

```release-note
ipam/crd: Fix spurious "Unable to update CiliumNode custom resource" failures in cilium-agent
```

twpayne commented on pkg/ipam/crd.go:
```diff
@@ -128,12 +128,8 @@ func newNodeStore(nodeName string, conf Configuration, owner Owner, k8sEventReg
-							equal = true
-							return
-						}
 						valid = true
```
It looks like there was another bug in the old code here: `valid` was only set to `true` if `oldNode` and `newNode` were not equal. This PR looks like it fixes that.

twpayne commented on pkg/ipam/crd.go:
```diff
@@ -128,12 +128,8 @@ func newNodeStore(nodeName string, conf Configuration, owner Owner, k8sEventReg
-						}
 						valid = true
-						store.updateLocalNodeResource(newNode.DeepCopy())
+						equal = !store.updateLocalNodeResource(newNode.DeepCopy())
```
Overall, I found this change quite hard to understand, but I think a bit of renaming would help. Specifically:
* The name of the function`updateLocalNodeResource` does not indicate that it returns a value, yet here we a receiving a `bool` and inverting it.
*`updateLocalNodeResource` returns a `bool` which it calls `changed`, but here we invert it with `!` and call it `equal`.
* At line 120 in this file, we also call `updateLocalNodeResource` but ignore its return value.

`updateLocalNodeResource` is really doing multiple things: storing the latest node value, determining whether the node's spec or status has changed, and updating ENI routes and allocation pool sizes. I think the code should be refactored to at least move the change detection out of `updateLocalNodeResource`.

gandro replied:
Good feedback, thanks! I looked at the code again and found that introducing a side-effect free setter instead makes the change much easier to understand.

Please re-review.

gandro committed:
ipam/crd: Fix spurious CiliumNode update status failures

When running in CRD-based IPAM modes (Alibaba, Azure, ENI, CRD), it is
possible to observe spurious "Unable to update CiliumNode custom
resource" failures in the cilium-agent.

The full error message is as follows: "Operation cannot be fulfilled on
ciliumnodes.cilium.io <node>: the object has been modified; please apply
your changes to the latest version and try again".

It means that the Kubernetes `UpdateStatus` call has failed because the
local `ObjectMeta.ResourceVersion` of submitted CiliumNode version is
out of date. In the presence of races, this error is expected and will
resolve itself once the agent receives a more recent version of the
object with the new resource version.

However, it is possible that the resource version of a `CiliumNode`
object is bumped even though the `Spec` or `Status` of the `CiliumNode`
remains the same. This for examples happens when
`ObjectMeta.ManagedFields` is updated by the Kubernetes apiserver.

Unfortunately, `CiliumNode.DeepEqual` does _not_ consider any
`ObjectMeta` fields (including the resource version). Therefore two
objects with different resource versions are considered the same by the
`CiliumNode` watcher used by IPAM.

But to be able to successfully call `UpdateStatus` we need to know the
most recent resource version. Otherwise, `UpdateStatus` will always fail
until the `CiliumNode` object is updated externally for some reason.

Therefore, this commit modifies the logic to always store the most
recent version of the `CiliumNode` object, even if `Spec` or `Status`
has not changed.  This in turn allows `nodeStore.refreshNode` (which
invokes `UpdateStatus`) to always work on the most recently observed
resource version.

Signed-off-by: Sebastian Wicki <sebastian@isovalent.com>

twpayne reviewed (approved):
This is great, thanks!
